lib/tcpip/network_wrapper.cc

```diff
@@ -590,6 +689,128 @@ SObj network_socket_create_and_bind(Timeout       *timeout,
+		  uint32_t                 addressLength = sizeof(addressTmp);
+		  auto                     rawSocket     = FreeRTOS_accept(
+		                            listeningSocket->socket, &addressTmp, &addressLength);
+		  if (rawSocket == nullptr)
+		  {
+			  Debug::log("Failed to create socket.");
+			  // This cannot fail unless buggy - we know that we
+			  // successfully allocated the token with this malloc
+			  // capability. Same for other calls to `token_obj_destroy`
+			  // in this function.
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+		  socketWrapper->socket = rawSocket;
+
+		  // Claim the socket so that it counts towards the caller's quota.  The
+		  // network stack also keeps a claim to it.  We will drop this claim on
+		  // deallocation.
+		  Claim c{mallocCapability, rawSocket};
+		  if (!c)
+		  {
+			  Debug::log("Failed to claim socket.");
+			  // Note that `close_socket_retry` can fail, in which case we
+			  // will leak the socket allocation. There is nothing we can do
+			  // here.  Returning the half-baked sealed socket would be
+			  // dangerous because the close() path isn't designed to handle
+			  // it (and changing it to do so is non-trivial and likely not
+			  // worth the trouble).
+			  close_socket_retry(timeout, rawSocket);
+			  // Cannot fail, see above.
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+
+		  if (LockGuard g{sealedSocketsListLock, timeout})
+		  {
+			  // Add the socket to the sealed socket reset list.
+			  socketWrapper->ring.cell_reset();
+			  sealedSockets.append(&(socketWrapper->ring));
+		  }
+		  else
+		  {
+			  // See above comments.
+			  Debug::log("Failed to add socket to the socket reset list.");
+			  close_socket_retry(timeout, rawSocket);
+			  token_obj_destroy(mallocCapability, socket_key(), sealedSocket);
+			  return -EINVAL;
+		  }
+
+		  // Set `address`.
+		  if ((heap_claim_fast(timeout, address) < 0) ||
+		      (!check_pointer<PermissionSet{Permission::Store}>(address)))
+		  {
+			  // There is not much we can do if this happens. The
+			  // caller passed an invalid `address` or freed it
+			  // concurrently: we can simply ignore, they are
+			  // shooting themselves in the foot by making
+			  // `address` unusable. Returning nullptr at that
+			  // stage would force us to close the socket, destroy
+			  // the token, etc.
+			  Debug::log("Invalid address pointer");
+		  }
+		  else
+		  {
+			  if (addressTmp.sin_family == FREERTOS_AF_INET6)
+			  {
+				  address->kind = NetworkAddress::AddressKindIPv6;
+				  memcpy(
+				    address->ipv6, addressTmp.sin_address.xIP_IPv6.ucBytes, 16);
+			  }
+			  else
+			  {
+				  address->kind = NetworkAddress::AddressKindIPv4;
+				  address->ipv4 = addressTmp.sin_address.ulIP_IPv4;
+			  }
+		  }
+		  // Set `port`.
+		  if ((heap_claim_fast(timeout, port) < 0) ||
```

Comment:
As above.

Reply:
Same here? I intended to claim fast the `port`.